File: cdk_python/serverless_stack.py
from aws_cdk import (core as cdk, aws_lambda, aws_apigateway, aws_dynamodb)


class ServerlessStack(cdk.Stack):

    def __init__(self, scope: cdk.Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Create Dynamo Table
        table_name = 'cdkPythonTable'
        table = aws_dynamodb.Table(self, table_name,
                                   table_name=table_name,
                                   partition_key=aws_dynamodb.Attribute(
                                       name='id',
                                       type=aws_dynamodb.AttributeType.STRING
                                   ))

        # Create lambda function
        function = aws_lambda.Function(self, 'LambdaFunction',
                                       handler='function.handler',
                                       function_name='simple-function',
                                       runtime=aws_lambda.Runtime.PYTHON_3_8,
                                       code=aws_lambda.Code.asset("./lambda"),
                                       environment={
                                           'DB_TABLE': table_name
                                       })

        # Grant function read on table
        table.grant_read_write_data(function)

        # Create API proxy Gateway
        api = aws_apigateway.LambdaRestApi(self, 'APIGateway',
                                           handler=function,
                                           rest_api_name='APIGateway',
                                           description='Rest API for lambda function')

        # Expose API's url output
        self.url_output = cdk.CfnOutput(self, 'ApiGatewayUrl',
                                        value=api.url)

File: src/ServerlessStack.ts
import * as lambda from '@aws-cdk/aws-lambda';
import * as api from '@aws-cdk/aws-apigateway';
import * as cdk from '@aws-cdk/core';
import { AttributeType, Table } from '@aws-cdk/aws-dynamodb';
import { CfnOutput } from '@aws-cdk/core';

export class ServerlessStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create DynamoDB
    const tableName = 'cdkTable'
    let dynamoTable = new Table(this, "CdkDynamoDBTable", {
      tableName,
      partitionKey: {name: 'id', type: AttributeType.STRING},
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });

    // Create Lambda function with an existing handler
    const fn = new lambda.Function(this, "HelloFn", {
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('./resource'),
      functionName: 'CdkHelloFn',
      environment: {
        DB_TABLE: dynamoTable.tableName
      }
    });
    
    // Grant read access to lambda function
    dynamoTable.grantReadData(fn)

    // Create API Gateway with GET method triggering Lambda function
    const gateway = new api.LambdaRestApi(this, "HelloFnApi", {
      handler: fn,
      restApiName: 'HelloFnApi',
      description: 'Api gateway created by CDK for HelloFn Lambda',
      proxy: false
    });

    gateway.root.resourceForPath('scan').addMethod('GET', new api.LambdaIntegration(fn));

    // Expose api url to output
    new CfnOutput(this, 'ScanApiGatewayUrl', {
      value: gateway.url,
      exportName: 'ScanApiGatewayUrl'
    })
  }
}
